File: sample-conference-videochat/sample-conference-videochat/Components/ChatScreen/Views/InputToolbar/KVOView.m
//
//  KVOView.m
//  sample-conference-videochat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "KVOView.h"


static void * kFrameKeyValueObservingContext = &kFrameKeyValueObservingContext;

@interface KVOView()

@property (assign, nonatomic, getter=isObserverAdded) BOOL observerAdded;

@end

@implementation KVOView

#pragma mark - Actions

- (void)setCollectionView:(UICollectionView *)collectionView {
    
    _collectionView = collectionView;
    
    if (floor(NSFoundationVersionNumber) >= NSFoundationVersionNumber_iOS_9_0) {
        [_collectionView.panGestureRecognizer addTarget:self
                                                 action:@selector(handlePanGestureRecognizer:)];
    }
}

- (void)willMoveToSuperview:(UIView *)newSuperview {
    
    if (self.isObserverAdded) {
        
        if (self.hostViewFrameChangeBlock) {
            self.hostViewFrameChangeBlock(newSuperview, NO);
        }
        
        [self.superview removeObserver:self
                            forKeyPath:@"center"
                               context:kFrameKeyValueObservingContext];
    }
    
    [newSuperview addObserver:self
                   forKeyPath:@"center"
                      options:NSKeyValueObservingOptionNew
                      context:kFrameKeyValueObservingContext];
    
    self.observerAdded = YES;
    
    [super willMoveToSuperview:newSuperview];
}

#pragma mark - Key-value observing

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context {
    
    if ([keyPath isEqualToString:@"center"] ) {
        if (self.hostViewFrameChangeBlock) {
            self.hostViewFrameChangeBlock(self.superview,
                                          _collectionView.panGestureRecognizer.state != UIGestureRecognizerStateChanged);
        }
    }
}

- (void)handlePanGestureRecognizer:(UIPanGestureRecognizer *)gesture {
    
    if (self.superview == nil) {
        return;
    }
    
    if (gesture.state == UIGestureRecognizerStateChanged) {
        
        UIView *host = self.superview;
        UIView *input = self.inputView;
        
        CGRect frame = host.frame;
        const CGPoint panPoint = [gesture locationInView:input.window];
        const CGRect hostViewRect = [input convertRect:frame toView:host];
        
        if (panPoint.y >= hostViewRect.origin.y) {
            frame.origin.y += hostViewRect.origin.y - panPoint.y;
            host.frame = frame;
        }
    }
}

@end

File: Swift-QMChatViewController/ChatViewController/ChatViewController/Views/KVOView.swift
//
//  KVOView.swift
//  Swift-ChatViewController
//
//  Created by Vladimir Nybozhinsky on 11/12/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

class KVOView: UIView {
  //MARK: - Properties
  var hostViewFrameChangeBlock : ((_ view: UIView, _ animated: Bool) -> Void)?
  
  var frameKeyValueObservingContext = UnsafeMutableRawPointer.allocate(byteCount: 4 * 4, alignment: 1)
  
  lazy var collectionView: UICollectionView = {
    let collectionView = UICollectionView()
    if floor(NSFoundationVersionNumber) >= NSFoundationVersionNumber_iOS_9_0 {
      collectionView.panGestureRecognizer.addTarget(self, action: #selector(handle(_:)))
    }
    return collectionView
  }()
  
  private var observerAdded: Bool = false {
    didSet {
      debugPrint("observerAdded \(observerAdded)")
    }
  }
  
  //MARK: - Life Cycle
  deinit {
    debugPrint("KVOView dealloc")
  }
  
  //MARK: - Overrides
  override func willMove(toSuperview newSuperview: UIView?) {
    
    if observerAdded {
      
      if let hostViewFrameChangeBlock = hostViewFrameChangeBlock,
        let newSuperview = newSuperview,
        let superview = superview {
        hostViewFrameChangeBlock(newSuperview, false)
        superview.removeObserver(self as NSObject, forKeyPath: "center",
                                 context: frameKeyValueObservingContext)
      }
    }
    newSuperview?.addObserver(self as NSObject, forKeyPath: "center", options: .new,
                              context: frameKeyValueObservingContext)
    observerAdded = true
    super.willMove(toSuperview: newSuperview)
  }
  
  // MARK: - Key-value observing
  override func observeValue(forKeyPath keyPath: String?, of object: Any?,
                             change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
    
    if (keyPath == "center") {
      if let hostViewFrameChangeBlock = hostViewFrameChangeBlock,
        let superview = superview {
        hostViewFrameChangeBlock(superview, collectionView.panGestureRecognizer.state != .changed)
      }
    }
  }
  
  //MARK: - Internal Methods
  @objc private func handle(_ gesture: UIPanGestureRecognizer) {
    
    if superview == nil {
      return
    }
    
    if gesture.state == .changed {
      
      guard let host = superview, let input = inputView else { return }
      
      var frame = host.frame
      let panPoint = gesture.location(in: input.window)
      let hostViewRect = input.convert(frame, to: host)
      
      if panPoint.y >= hostViewRect.origin.y {
        frame.origin.y += hostViewRect.origin.y - panPoint.y
        host.frame = frame
      }
    }
  }
}
